mlflow/store/artifact/databricks_artifact_repo.py

```diff
@@ -0,0 +1,94 @@
+from azure.storage.blob import BlobClient
+
+import os
+from mlflow.exceptions import MlflowException
+from mlflow.store.artifact.artifact_repo import ArtifactRepository
+from mlflow.utils.string_utils import strip_suffix
+from mlflow.utils.file_utils import relative_path_to_artifact_path
+from mlflow.utils.rest_utils import call_endpoint, extract_api_info_for_service
+from mlflow.protos.databricks_artifacts_pb2 import DatabricksMlflowArtifactsService
+from mlflow.protos.databricks_artifacts_pb2 import GetCredentialsForWrite, GetCredentialsForRead
+from mlflow.utils.databricks_utils import get_databricks_host_creds
+from mlflow.protos.service_pb2 import MlflowService, ListArtifacts
+
+_PATH_PREFIX = "/api/2.0"
+
+
+class DatabricksArtifactRepository(ArtifactRepository):
+    """
+    SOMETHING : TYPING TILL IT WORKS LOL
+    """
+
+    def __init__(self, artifact_uri):
+        super(DatabricksArtifactRepository, self).__init__(artifact_uri)
+
+    def _extract_run_id(self, artifact_uri):
+        return artifact_uri.lstrip('/').split('/')[4]
```

Comment:
@arjundc-db Awesome that we're extracting the run ID from the URI! A quick suggestion here - can we: 1. use the urlparse library to fetch the path component of the URI and 2. normalize the path before performing the splitting? E.g.,
```
import posixpath
from urllib.parse import urlparse
parsed_uri = urlparse(artifact_uri)
parsed_path = posixpath.normpath(parsed_uri.path)
return parsed_path.lstrip("/").split("/")[4]
```

The reasoning for #1 here is that DBFS URIs support empty hostnames; e.g., `dbfs:///databricks/mlflow-tracking/EXP_ID/RUN_ID` is valid (note the 3 leading slashes). I confirmed this by running `%fs ls dbfs:///databricks` in a Databricks notebook. The current solution yields the incorrect result when applied directly to a URI of this form:
```
In [1]: url = "dbfs:///databricks/mlflow-tracking/EXP_ID/RUN_ID/artifacts"
In [2]: url.lstrip('/').split('/')[4]
Out[2]: 'mlflow-tracking' # Should be 'RUN_ID'
```

Similarly, I've confirmed that `dbfs://databricks/` (two leading slashes) and `dbfs:databricks/...` (no leading slashes) are *invalid* use cases; Databricks rejects URIs of the first form because they contain non-empty hostnames, which are unsupported. Databricks rejects URIs of the second form because they contain *relative* paths. Databricks requires URIs to specify absolute DBFS paths. This means that we don't have to worry about relative paths, which may not contain a leading slash.

The reasoning for #2 here is that URI paths may contain redundant slashes. For example, `dbfs:/databricks/mlflow-tracking////EXP_ID/RUN_ID` is a valid URI that is semantically equivalent to `dbfs:/databricks/mlflow-tracking/EXP_ID/RUN_ID`. Normalizing the path will remove redundant slashes. Without normalizing the path, we currently get an incorrect result:

```
In [1]: url = "dbfs:/databricks////mlflow-tracking/EXP_ID/RUN_ID/"
In [2]: url.lstrip('/').split('/')[4]
Out [2]: ''
```

For more information about the structure of URIs, https://en.wikipedia.org/wiki/Uniform_Resource_Identifier is an awesome reference.  Let me know if you have any questions here.

**These kinds of URI edge cases would be great to handle via a unit test. Let's make sure we plan to add a unit test for this case before merging this functionality!**

Reply:
Once we correct the behavior here, can we also add an inline comment that clearly documents the assumptions we're making about the URI structure? E.g., URIs are assumed to be semantically equivalent to `dbfs:/databricks/mlflow-tracking/EXP_ID/RUN_ID/artifacts/...` ? This will help future readers of the code interpret the behavior.

Reply:
Done, this pretty cool to read about, thanks!